File: rust/examples/mat_vec_mul.rs
#![no_std]
#![no_main]
#![allow(unused)]

// Crates being used
// =================

use riscv_rt::entry;

extern crate nocl;
use nocl::*;
use nocl::rand::*;
use nocl::prims::*;

extern crate alloc;
use alloc::vec;
use alloc::vec::*;
use alloc::boxed::*;

// Benchmark
// =========

const BLOCK_SIZE: usize = prims::config::SIMT_LANES as usize;

struct MatVecMul {
  width   : usize,
  height  : usize,
  mat     : Buffer<i32>,
  vec_in  : Buffer<i32>,
  vec_out : Buffer<i32>
}

impl Code for MatVecMul {

#[inline(always)]
fn run (my : &My, shared : &mut Scratch, params: &mut MatVecMul) {
  let mut partial = alloc::<i32>(shared, BLOCK_SIZE);

  for y in (my.block_idx.x .. params.height).step_by(my.grid_dim.x) {
    // Row processed by this block
    let row = y * params.width;

    // Compute partial dot products
    let mut sum = 0;
    for x in (my.thread_idx.x .. params.width).step_by(my.block_dim.x) {
      sum += params.mat[row+x] * params.vec_in[x];
    }
    partial[my.thread_idx.x] = sum;
    syncthreads();

    // Final local reduction
    let mut i = my.block_dim.x >> 1;
    while i > 0 {
      if my.thread_idx.x < i {
        partial[my.thread_idx.x] += partial[my.thread_idx.x + i];
      }
      syncthreads();
      i = i >> 1
    }

    // Write dot product to global memory
    if my.thread_idx.x == 0 { params.vec_out[y] = partial[0] }

    nocl_converge()
  }
}

}

#[entry]
fn main() -> ! {
  nocl_init();

  // Matrix size for benchmarking
  #[cfg(not(feature = "large_data_set"))]
  const WIDTH : usize = 128;
  #[cfg(not(feature = "large_data_set"))]
  const HEIGHT : usize = 64;
  #[cfg(feature = "large_data_set")]
  const WIDTH : usize = 1024;
  #[cfg(feature = "large_data_set")]
  const HEIGHT : usize = 1024;

  // Input and output matrix data
  let mut mat : Vec<i32> = vec![0; WIDTH*HEIGHT];
  let mut vec_in : Vec<i32> = vec![0; WIDTH];
  let mut vec_out : Vec<i32> = vec![0; HEIGHT];

  // Initialise inputs
  let mut seed : u32 = 1;
  for j in 0..WIDTH {
    vec_in[j] = rand15(&mut seed) as i32
  }
  for i in 0..HEIGHT {
    for j in 0..WIDTH {
      mat[i*WIDTH + j] = rand15(&mut seed) as i32
    }
  }

  // Block/grid dimensions
  let dims =
    Dims {
      block_dim: Dim2 { x: prims::config::SIMT_LANES,
                        y: 1 },
      grid_dim: Dim2 { x: prims::config::SIMT_WARPS,
                       y: 1 }
    };
 
  // Kernel parameters
  let mut params =
    MatVecMul { width   : WIDTH,
                height  : HEIGHT,
                mat     : mat.into(),
                vec_in  : vec_in.into(),
                vec_out : vec_out.into() };

  // Invoke kernel
  let params = nocl_run_kernel_verbose(dims, params);

  // Check result
  let mut ok = true;
  for i in 0 .. HEIGHT {
    let mut sum = 0;
    for j in 0 .. WIDTH {
      sum += params.mat[i*WIDTH+j] * params.vec_in[j]
    }
    ok = ok && sum == params.vec_out[i]
  }

  // Display result
  putstr("Self test: ");
  putstr(if ok { "PASSED" } else { "FAILED" });
  putchar(b'\n');

  // The end
  end()
}

File: apps/Samples/MatVecMul/MatVecMul.cpp
#include <NoCL.h>
#include <Rand.h>

// Kernel for matrix-vector multipliation
template <int BlockSize> struct MatVecMul : Kernel {
  int width, height;
  int *mat, *vecIn, *vecOut;
 
  // Partial dot products stored in shared local memory
  int* partial;

  INLINE void init() {
    declareShared(&partial, BlockSize);
  }
 
  INLINE void kernel() {
    for (int y = blockIdx.x; y < height; y += gridDim.x) {
      // Row processed by this block
      int* row = mat + y * width;

      // Compute partial dot products
      int sum = 0;
      for (int x = threadIdx.x; x < width; x += blockDim.x)
        sum += row[x] * vecIn[x];
      partial[threadIdx.x] = sum;
      __syncthreads();

      // Final local reduction
      for (int i = blockDim.x >> 1; i > 0; i >>= 1)  {
        if (threadIdx.x < i)
          partial[threadIdx.x] += partial[threadIdx.x + i];
        __syncthreads();
      }

      // Write dot product to global memory
      if (threadIdx.x == 0) vecOut[y] = partial[0];

      noclConverge();
    }
  }
};

int main()
{
  // Are we in simulation?
  bool isSim = getchar();

  // Vector and matrix dimensions for benchmarking
  int width = isSim ? 128 : 1024;
  int height = isSim ? 64 : 1024;

  // Input and outputs
  simt_aligned int mat[height*width], vecIn[width], vecOut[height];

  // Initialise inputs
  uint32_t seed = 1;
  for (int j = 0; j < width; j++)
    vecIn[j] = rand15(&seed) & 0xff;
  for (int i = 0; i < height; i++) {
    for (int j = 0; j < width; j++)
      mat[i*width+j] = rand15(&seed) & 0xff;
  }

  // Instantiate kernel
  MatVecMul<SIMTLanes> k;

  // One block of threads per matrix row
  k.blockDim.x = SIMTLanes;
  k.gridDim.x = SIMTWarps;

  // Assign parameters
  k.width = width;
  k.height = height;
  k.mat = mat;
  k.vecIn = vecIn;
  k.vecOut = vecOut;

  // Invoke kernel
  noclRunKernelAndDumpStats(&k);

  // Check result
  bool ok = true;
  for (int i = 0; i < height; i++) {
    int sum = 0;
    for (int j = 0; j < width; j++)
      sum += mat[i*width+j] * vecIn[j];
    ok = ok && sum == vecOut[i];
  }

  // Display result
  puts("Self test: ");
  puts(ok ? "PASSED" : "FAILED");
  putchar('\n');

  return 0;
}
